### PR TITLE
Vectorize TensorPrimitives.Acosh/sinh/tanh in terms of Log

### PR DESCRIPTION
**Pull request by stephentoub**
Contributes to https://github.com/dotnet/runtime/issues/97193

### CODE REVIEW

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs [hunk 1]
```diff
@@ -15868,11 +15868,54 @@ public static Vector512<T> Invoke(Vector512<T> x)
...
-            public static Vector128<T> Invoke(Vector128<T> x) => throw new NotSupportedException();
-            public static Vector256<T> Invoke(Vector256<T> x) => throw new NotSupportedException();
-            public static Vector512<T> Invoke(Vector512<T> x) => throw new NotSupportedException();
+
+            public static Vector128<T> Invoke(Vector128<T> x)
+            {
+                if (typeof(T) == typeof(float))
+                {
+                    Vector128<float> f = x.AsSingle();
+                    return LogOperator<float>.Invoke(f + Vector128.Sqrt(MultiplyAddEstimateOperator<float>.Invoke(f, f, Vector128.Create(-1f)))).As<float, T>();
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs:15880]: Isn't it supposed to be `log(f + sqrt(f - 1) * sqrt(f + 1))`?

Comment (tannergooding) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs:15880]: Doing `sqrt(f * f - 1)` only works assuming `f` is positive, which we can't make here without doing `abs` and `copysign`

File: src/libraries/System.Numerics.Tensors/tests/TensorPrimitives.Generic.cs [hunk 2]
```diff
@@ -365,7 +365,7 @@ public static IEnumerable<object[]> SpanDestinationFunctionsToTest()
             yield return new object[] { new SpanDestinationDelegate(TensorPrimitives.Acosh), new Func<T, T>(T.Acosh) };
             yield return new object[] { new SpanDestinationDelegate(TensorPrimitives.AcosPi), new Func<T, T>(T.AcosPi) };
             yield return new object[] { new SpanDestinationDelegate(TensorPrimitives.Acos), new Func<T, T>(T.Acos) };
-            yield return new object[] { new SpanDestinationDelegate(TensorPrimitives.Asinh), new Func<T, T>(T.Asinh) };
+            yield return new object[] { new SpanDestinationDelegate(TensorPrimitives.Asinh), new Func<T, T>(T.Asinh), T.CreateTruncating(0.001) };
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Tensors/tests/TensorPrimitives.Generic.cs:368]: This is quite a lot of variance being allowed, we're basically saying we can't get within 3 digits of accuracy for some inputs.

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs:15880]: > Isn't it supposed to be log(f + sqrt(f - 1) * sqrt(f + 1))?

Why? Isn't the formula `acosh(x) = ln(x + sqrt(x^2 - 1))`?

> we need to ensure that invalid inputs still return NaN and that acos(1) returns exactly +0

It does, e.g. this:
```C#
using System.Numerics.Tensors;

float[] values = new float[] { 0f, 1f, 2f, 3f };
float[] dest = new float[values.Length];
TensorPrimitives.Acosh<float>(values, dest);

Console.WriteLine(float.IsNaN(dest[0]));
Console.WriteLine(BitConverter.SingleToUInt32Bits(dest[1]));
Console.WriteLine(BitConverter.SingleToUInt32Bits(dest[2]));
Console.WriteLine(BitConverter.SingleToUInt32Bits(dest[3]));
```
outputs:
```
True
0
1068012052
1071751603
```

> Doing sqrt(f * f - 1) only works assuming f is positive

Why does it assume that?

File: src/libraries/System.Numerics.Tensors/tests/TensorPrimitives.Generic.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Numerics.Tensors/tests/TensorPrimitives.Generic.cs:368]: What do you suggest?

(Note that the default being used everywhere right now is 0.0001.)

File: src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs [hunk 3]
```diff
@@ -15934,11 +16021,54 @@ public static Vector512<T> Invoke(Vector512<T> x)
...
-            public static Vector128<T> Invoke(Vector128<T> x) => throw new NotSupportedException();
-            public static Vector256<T> Invoke(Vector256<T> x) => throw new NotSupportedException();
-            public static Vector512<T> Invoke(Vector512<T> x) => throw new NotSupportedException();
+
+            public static Vector128<T> Invoke(Vector128<T> x)
+            {
+                if (typeof(T) == typeof(float))
+                {
+                    Vector128<float> f = x.AsSingle();
+                    return (Vector128.Create(0.5f) * LogOperator<float>.Invoke((Vector128<float>.One + f) / (Vector128<float>.One - f))).As<float, T>();
```
Comment (lilinus) [inline src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs:16033]: A more numerically stable formula for small f would be to use lopg1
`Vector128.Create(0.5f) * LogP1Operator<float>.Invoke(Vector128<float>.Create(2f) * f / (Vector128<float>.One - f))`

**Comment by stephentoub** [comment]
This was a background activity for me and I'd intended for it to be a background activity for others, but I'm told it's a distraction, so I'm going to close it for now and it can be resuscitated at some point in the future. Thanks.

**Pull request closed without merging**